Logging overhaul, fmt::format, and related

jagerman opened the pull request:
This replaces custom logging system with spdlog-based oxen logging.  This PR mainly replaces the backend logging with the spdlog-based system, but doesn't (yet) convert all the existing LogWarn, etc. to use the new format-based logging.

New logging statements will look like:

    llarp::log::warning(cat, "blah: {}", val);

where `cat` should be set up in each .cpp or cluster of .cpp files, as described in the [oxen-logging README](https://github.com/oxen-io/oxen-logging)

As part of spdlog we get fmt, which gives us nice format strings, where are applied generously in this PR.

Making types printable now requires two steps:
- add a ToString() method
- add this specialization:

      template <>
      constexpr inline bool llarp::IsToStringFormattable<llarp::Whatever> = true;

This will then allow the type to be printed as a "{}" value in a fmt::format string.  This is applied to all our printable types here, and all of the `operator<<` are removed.

This commit also:
- replaces various uses of `operator<<` to ToString()
- replaces various uses of std::stringstream with either fmt::format or plain std::string
- Rename some to_string and toString() methods to ToString() for consistency (and to work with fmt)
- Replace `stringify(...)` and `make_exception` usage with fmt::format (and remove stringify/make_exception from util/str.hpp).

jagerman committed:
Add missing header

Needed for uint_least32_t.

jagerman committed:
Replace logging with oxen-logger

Replaces custom logging system with spdlog-based oxen logging.  This
commit mainly replaces the backend logging with the spdlog-based system,
but doesn't (yet) convert all the existing LogWarn, etc. to use the new
format-based logging.

New logging statements will look like:

    llarp::log::warning(cat, "blah: {}", val);

where `cat` should be set up in each .cpp or cluster of .cpp files, as
described in the oxen-logging README.

As part of spdlog we get fmt, which gives us nice format strings, where
are applied generously in this commit.

Making types printable now requires two steps:
- add a ToString() method
- add this specialization:

      template <>
      constexpr inline bool llarp::IsToStringFormattable<llarp::Whatever> = true;

This will then allow the type to be printed as a "{}" value in a
fmt::format string.  This is applied to all our printable types here,
and all of the `operator<<` are removed.

This commit also:
- replaces various uses of `operator<<` to ToString()
- replaces various uses of std::stringstream with either fmt::format or
  plain std::string
- Rename some to_string and toString() methods to ToString() for
  consistency (and to work with fmt)
- Replace `stringify(...)` and `make_exception` usage with fmt::format
  (and remove stringify/make_exception from util/str.hpp).

jagerman commented:
Still todo:
- [x] macos support
- [x] get rid of the `.print(blah, -1, -1)` print layer

majestrate commented on llarp/config/config.cpp:
```diff
@@ -1097,26 +1100,27 @@ namespace llarp
-          LogType type = LogTypeFromString(arg);
-          if (type == LogType::Unknown)
-            throw std::invalid_argument(stringify("invalid log type: ", arg));
+          auto type = log::type_from_string(arg);
```
`log::type_from_string` should throw imo

jagerman replied:
It throws now (as does `level_from_string`).

majestrate commented on llarp/dns/message.hpp:
```diff
@@ -96,19 +96,23 @@ namespace llarp
-      return out;
-    }
   }  // namespace dns
+
```
should be outside the llarp namespace

majestrate commented on llarp/net/address_info.hpp:
```diff
@@ -55,23 +55,23 @@ namespace llarp
   bool
   operator<(const AddressInfo& lhs, const AddressInfo& rhs);
 
+  template <>
```
move out of llarp namespace?

majestrate commented on llarp/net/sock_addr.hpp:
```diff
@@ -60,9 +61,9 @@ namespace llarp
-    operator const sockaddr_in*() const;
-    operator const sockaddr_in6*() const;
+    explicit operator const sockaddr*() const;
+    explicit operator const sockaddr_in*() const;
```
i like this change

majestrate commented on llarp/service/info.hpp:
```diff
@@ -101,11 +104,8 @@ namespace llarp
 }  // namespace llarp
+
+template <>
+constexpr inline bool llarp::IsToStringFormattable<llarp::service::ServiceInfo> = true;
```
wait... so does this template specialization belong inside or outside the llarp namespace?

jagerman committed:
Make test code work with new logging

jagerman committed:
Avoid ctor inheritance to make diagnostics better

Using constructor inheritance DRYs the code, but unfortunately confuses
GCC as to where the proper "required from here" location is, which makes
debugging formatting errors very difficult.  Avoid it (and update
oxen-logging to avoid it there as well).

jagerman committed:
Remote util::Printer and related cruft

jagerman committed:
Use more fmt

jagerman committed:
Add missing header to fix libc++ build

jagerman committed:
gcc 8/9 fix

majestrate committed:
fix win32 and android builds

jagerman committed:
Update oxen-logging for macos compilation

jagerman committed:
oxen-logging update to handle level/type parsing exceptions

jagerman committed:
hive fmt/spdlog updates
